### PR TITLE
Support defining types in modules (BREAKING)

### LINKED ISSUE #930: Support defining types in modules
Currently, all Carp types are global which can lead to odd behavior when defining them in modules. As a basic illustration of this behavior, consider the following example:

```clojure
(defmodule Homotypic (deftype (Pair a a) [fst a snd a]))
鲤 :i Pair
Homotypic.Pair : Type
Defined at line 1, column 32 in 'REPL'
Pair : Module = { ;; seems to just be the original Pair from core. 
    < : (Fn [(Pair a b), (Pair a b)] Bool)
    = : (Fn [(Pair a b), (Pair a b)] Bool)
    > : (Fn [(Pair a b), (Pair a b)] Bool)
    a : (Fn [(Ref (Pair a b) q)] (Ref a q))
    b : (Fn [(Ref (Pair a b) q)] (Ref b q))
    copy : (Fn [(Ref (Pair a b) q)] (Pair a b))
    delete : (Fn [(Pair a b)] ())
    hash : (Fn [(Ref (Pair a b) c)] Int)
    init : (Fn [a, b] (Pair a b))
    init-from-refs : (Fn [(Ref a b), (Ref c d)] (Pair a c))
    prn : (Fn [(Ref (Pair a b) q)] String)
    reverse : (Fn [(Ref (Pair a b) c)] (Pair b a))
    set-a : (Fn [(Pair a b), a] (Pair a b))
    set-a! : (Fn [(Ref (Pair a b) q), a] ())
    set-b : (Fn [(Pair a b), b] (Pair a b))
    set-b! : (Fn [(Ref (Pair a b) q), b] ())
    str : (Fn [(Ref (Pair a b) q)] String)
    update-a : (Fn [(Pair a b), (Ref (Fn [a] a fq) q)] (Pair a b))
    update-b : (Fn [(Pair a b), (Ref (Fn [b] b fq) q)] (Pair a b))
    zero : (Fn [] (Pair a b))
}
Defined at line 91, column 11 in '/Users/scottolsen/dev/Carp/core/Tuples.carp'
Documentation: is a 2-tuple, i.e. a datatype with two members.
鲤 :i Homotypic.Pair
Pair : Module = { ;; original Pair from core + Homotypic Pair?!
    < : (Fn [(Pair a b), (Pair a b)] Bool)
    = : (Fn [(Pair a b), (Pair a b)] Bool)
    > : (Fn [(Pair a b), (Pair a b)] Bool)
    a : (Fn [(Ref (Pair a b) q)] (Ref a q))
    b : (Fn [(Ref (Pair a b) q)] (Ref b q))
    copy : (Fn [(Ref (Pair a a) q)] (Pair a a))
    delete : (Fn [(Pair a a)] ())
    fst : (Fn [(Ref (Pair a a) q)] (Ref a q))
    hash : (Fn [(Ref (Pair a b) c)] Int)
    init : (Fn [a, a] (Pair a a))
    init-from-refs : (Fn [(Ref a b), (Ref c d)] (Pair a c))
    prn : (Fn [(Ref (Pair a a) q)] String)
    reverse : (Fn [(Ref (Pair a b) c)] (Pair b a))
    set-a : (Fn [(Pair a b), a] (Pair a b))
    set-a! : (Fn [(Ref (Pair a b) q), a] ())
    set-b : (Fn [(Pair a b), b] (Pair a b))
    set-b! : (Fn [(Ref (Pair a b) q), b] ())
    set-fst : (Fn [(Pair a a), a] (Pair a a))
    set-fst! : (Fn [(Ref (Pair a a) q), a] ())
    set-snd : (Fn [(Pair a a), a] (Pair a a))
    set-snd! : (Fn [(Ref (Pair a a) q), a] ())
    snd : (Fn [(Ref (Pair a a) q)] (Ref a q))
    str : (Fn [(Ref (Pair a a) q)] String)
    update-a : (Fn [(Pair a b), (Ref (Fn [a] a fq) q)] (Pair a b))
    update-b : (Fn [(Pair a b), (Ref (Fn [b] b fq) q)] (Pair a b))
    update-fst : (Fn [(Pair a a), (Ref (Fn [a] a fq) q)] (Pair a a))
    update-snd : (Fn [(Pair a a), (Ref (Fn [a] a fq) q)] (Pair a a))
    zero : (Fn [] (Pair a b))
}
```

This prevents library designers from using type names that are already declared in the standard library. While this isn't a major issue as one can always prefix a type name `(deftype HomotypicPair)` the experience might be nice if we allow for module resolution at the level of type as well, such that one can use shorted, more appropriate names locally and write, e.g.`(definterface foo (Fn [] (Homotypic.Pair)))` instead of having to attach prefixes to the type names themselves.

### PR DESCRIPTION
**Pull request by scolsen**
This commit adds support for defining types (using deftype) in modules.
Previously, all types were hoisted to the top level of the type
environment. After this commit, the type environment supports defining
nested modules just like the value env, so, calling the following:

```
(defmodule Foo (deftype Bar Baz))
```

Adds the following to the type env:

```
Foo : Module = {
    Bar : Type
}
```

and the following to the value env:

```
Foo : Module = {
    Bar : Module = {
        Baz : (Fn [] Foo.Bar)
        copy : (Fn [(Ref Foo.Bar q)] Foo.Bar)
        delete : (Fn [Foo.Bar] ())
        get-tag : (Fn [(Ref Foo.Bar q)] Int)
        prn : (Fn [(Ref Foo.Bar q)] String)
        str : (Fn [(Ref Foo.Bar q)] String)
    }
}

```

Such a type is *distinct* from any type defined at the top level that
happens to also have the name `Bar`.

This commit also updates info and tests to account for types in modules.

BREAKING CHANGE: This change is breaking since it alters the names of
types that were previously defined in modules. A good example of this is
the `Id` type in the `Color` module. Previously, one could refer to this
type by simply typing `Id` since it was hoisted to the top level. Now it
*must* be referred to by `Color.Id` since `Id` at the top level of the
type env and `Color.Id` (Id in the color module) are considered to be
distinct types.

fixes #930

### CODE REVIEW

**Comment by scolsen** [comment]
NOTE: This might break SDL--the machine I ran tests on doesn't have a copy of SDL, so I couldn't check them. If any of the SDL types are defined in a module it will break.

File: src/Scoring.hs [hunk 1]
```diff
@@ -17,36 +18,39 @@ scoreTypeBinder typeEnv b@(Binder _ (XObj (Lst (XObj x _ _ : XObj (Sym _ _) _ _
...
 
 depthOfDeftype :: TypeEnv -> Set.Set Ty -> XObj -> [Ty] -> Int
-depthOfDeftype typeEnv visited (XObj (Lst (_ : XObj (Sym (SymPath _ selfName) _) _ _ : rest)) _ _) varTys =
+depthOfDeftype typeEnv visited (XObj (Lst (_ : XObj (Sym (SymPath path selfName) _) _ _ : rest)) _ _) varTys =
   case concatMap expandCase rest of
     [] -> 100
     xs -> maximum xs
   where
-    depthsFromVarTys = map (depthOfType typeEnv visited selfName) varTys
+    depthsFromVarTys = map (depthOfType typeEnv visited (intercalate "" (path ++ [selfName]))) varTys
```
Comment (jacereda) [inline src/Scoring.hs:46]: `concat`?

File: src/Primitives.hs [hunk 2]
```diff
@@ -303,9 +304,11 @@ primitiveInfo _ ctx [target@(XObj (Sym path@(SymPath _ _) _) _ _)] = do
...
-      case lookupBinderInContextEnv ctx path of
-        Nothing -> notFound ctx target path
-        Just found -> printer found
+      (printIfFound (lookupBinderInTypeEnv ctx path))
+        >> ( case lookupBinderInContextEnv ctx path of
+               Nothing -> notFound ctx target path
+               Just found -> printer found
+           )
   where
     printIfFound :: Maybe Binder -> IO (Context, Either EvalError XObj)
```
Comment (jacereda) [inline src/Primitives.hs:313]: Perhaps not related to this PR, but shouldn't this return `IO ()`?

File: src/Primitives.hs [hunk 3]
```diff
@@ -676,7 +683,7 @@ primitiveDeftype _ _ _ = error "primitivedeftype"
 
 primitiveUse :: Primitive
 primitiveUse xobj ctx [XObj (Sym path _) _ _] =
-  pure $ maybe lookupInGlobal useModule (lookupInEnv path e)
+  (pure $ maybe lookupInGlobal useModule (lookupInEnv path e))
```
Comment (jacereda) [inline src/Primitives.hs:686]: The excess parenthesis will probably cause a warning in `HLint`

**Review by jacereda — COMMENTED** [review]
Nice!

**Comment by eriksvedang** [comment]
Woo, Christmas is coming early!

File: src/Types.hs [hunk 4]
```diff
@@ -328,3 +334,19 @@ typesDeleterFunctionType memberType = FuncTy [memberType] UnitTy StaticLifetimeT
...
+
+createStructName :: [String] -> String -> String
+createStructName path name = (intercalate "." (path ++ [name]))
+
+getStructName :: Ty -> String
+getStructName (StructTy (ConcreteNameTy name) _) = name
+getStructName (StructTy (VarTy name) _) = name
+getStructName _ = ""
+
+getPathFromStructName :: String -> [String]
```
Comment (eriksvedang) [inline src/Types.hs:346]: Does this mean that the struct type is stored with it's enclosing environments as part of the name for the type?

(e.g. `(defmodule A (defmodule B (deftype T)))` would create a type with the name "A.B.T")

**Comment by eriksvedang** [comment]
Got some conflicts now that I merged #1061 , I hope it's nothing too tricky. Let me know if you have any questions about the new code.

File: src/Types.hs (hunk 4 above)
Comment (scolsen) [reply src/Types.hs:346]: Yep, that's precisely correct.

I think ideally we wouldn't need to do this, but right now there are some functions that take a struct name as an argument and *then* use the name to perform an additional lookup--(see Scoring.hs) in particular. Possibly these can be refactored and eventually the name of the type can drop the module prefixes and we can rely on lookups alone to disambiguate.

File: src/Scoring.hs (hunk 1 above)
Comment (scolsen) [reply src/Scoring.hs]: thanks! bb777d4

File: src/Primitives.hs (hunk 3 above)
Comment (scolsen) [reply src/Primitives.hs]: 4bbc6bc

File: src/Primitives.hs (hunk 2 above)
Comment (scolsen) [reply src/Primitives.hs:315]: Definitely! Added a TODO about it: d963569

File: src/Primitives.hs [hunk 5]
```diff
@@ -642,10 +646,14 @@ primitiveDeftype xobj ctx (name : rest) =
                       )
                       i
                       (Just TypeTy)
+                  holderEnv = \name' prev -> Env (Map.fromList []) (Just prev) (Just name') [] ExternalEnv 0
```
Comment (eriksvedang) [inline src/Primitives.hs:649]: What does "holder" mean here?

File: src/Types.hs (hunk 4 above)
Comment (eriksvedang) [reply src/Types.hs:346]: Seems like `ConcreteNameTy` maybe should contain a `SymPath`..?

File: src/TypesToC.hs [hunk 6]
```diff
@@ -27,28 +29,29 @@ tyToCManglePtr :: Bool -> Ty -> String
...
+    f ByteTy = "uint8_t"
+    f StringTy = "String"
+    f PatternTy = "Pattern"
+    f CharTy = "Char"
+    f UnitTy = "void"
+    f (VarTy x) = x
+    f (FuncTy argTys retTy _) = "Fn__" ++ joinWithUnderscore (map (tyToCManglePtr True) argTys) ++ "_" ++ tyToCManglePtr True retTy
+    f (StructTy s []) = tyToCManglePtr False s
+    f (StructTy s typeArgs) = tyToCManglePtr False s ++ "__" ++ joinWithUnderscore (map (tyToCManglePtr True) typeArgs)
+    f (ConcreteNameTy name) = mangle (intercalate "" (map unpack (splitOn (pack ".") (pack name))))
```
Comment (eriksvedang) [inline src/TypesToC.hs:47]: `concat` here too?

File: src/Types.hs (hunk 4 above)
Comment (scolsen) [reply src/Types.hs:346]: Yes, that seems reasonable!

**Pull request merged**